### PR TITLE
enchance(usability delete flow cell command)

### PR DESCRIPTION
**Pull request by ChrOertlin**
## Description

Improvements to the delete flow cell command

### Added

-

### Changed

-

### Fixed

-

### How to prepare for test

- [ ] Ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_main -t cg -b [THIS-BRANCH-NAME] -a
    ```

### How to test

- [ ] Do ...

### Expected test outcome

- [ ] Check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [x] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

File: tests/meta/demultiplex/conftest.py [hunk 1]
```diff
@@ -83,6 +83,18 @@ def fixture_tmp_flow_cell_run_base_path(
     return tmp_flow_cell_run_path
 
 
+@pytest.fixture(name="tmp_flow_cell_demux_base_path")
+def fixture_tmp_flow_cell_demux_base_path(
+    project_dir: Path, bcl2fastq_flow_cell_full_name: str
+) -> Path:
+    """Flow cell run directory in temporary folder."""
```
Comment (diitaz93) [inline tests/meta/demultiplex/conftest.py:90]:
```suggestion
    """Flow cell demultiplex directory in temporary folder."""

```

File: tests/meta/demultiplex/conftest.py [hunk 2]
```diff
@@ -83,6 +83,18 @@ def fixture_tmp_flow_cell_run_base_path(
...
 
 
+@pytest.fixture(name="tmp_flow_cell_demux_base_path")
+def fixture_tmp_flow_cell_demux_base_path(
+    project_dir: Path, bcl2fastq_flow_cell_full_name: str
+) -> Path:
+    """Flow cell run directory in temporary folder."""
+
+    tmp_flow_cell_run_path: Path = Path(project_dir, "demultiplexing-runs")
+    tmp_flow_cell_run_path.mkdir(exist_ok=True, parents=True)
```
Comment (diitaz93) [inline tests/meta/demultiplex/conftest.py:93]:
```suggestion
    tmp_flow_cell_demux_path: Path = Path(project_dir, "demultiplexing-runs")
    tmp_flow_cell_demux_path.mkdir(exist_ok=True, parents=True)

```

File: tests/meta/demultiplex/conftest.py [hunk 3]
```diff
@@ -83,6 +83,18 @@ def fixture_tmp_flow_cell_run_base_path(
...
+
+    return tmp_flow_cell_run_path
```
Comment (diitaz93) [inline tests/meta/demultiplex/conftest.py:95]: Flow cell run is usually associated with the directory before demultiplexing
```suggestion
    return tmp_flow_cell_demux_path

```

**Comment by ChrOertlin** [comment]
```
 cg demultiplex delete-flow-cell -f someFC --demultiplexing-dir --run-dir
/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/sqlalchemy/orm/relationships.py:2231: SAWarning: Cascade settings "delete, merge, save-update" should not be combined with a viewonly=True relationship.   This configuration will raise an error in version 1.4.  Note that in versions prior to 1.4, these cascade settings may still produce a mutating effect even though this relationship is marked as viewonly=True.
  "viewonly=True." % (", ".join(sorted(non_viewonly)))
Running cg demultiplex.
Called undefined __fields__ on HousekeeperAPI, please wrap
Set environment to stage
Called undefined __fields__ on HousekeeperAPI, please wrap
Instantiating cgstats api
Are you sure you want to delete the flow cell from the following databases:
cg-stats=False
Demultiplexing-dir=True
Housekeeper=False
Init_files=False
Run-dir=True
Statusdb=False [y/N]: y
/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/sqlalchemy/orm/relationships.py:1998: SAWarning: Setting backref / back_populates on relationship Demux.samples to refer to viewonly relationship Sample.demuxes should include sync_backref=False set on the Demux.samples relationship.  (this warning may be suppressed after 10 occurrences)
  (rel_b, rel_a, rel_b),
/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/sqlalchemy/orm/relationships.py:1998: SAWarning: Setting backref / back_populates on relationship Sample.demuxes to refer to viewonly relationship Demux.samples should include sync_backref=False set on the Sample.demuxes relationship.  (this warning may be suppressed after 10 occurrences)
  (rel_b, rel_a, rel_b),
DeleteDemuxAPI-Hasta: Removing flow cell demultiplexing directory: /home/proj/stage/demultiplexed-runs/test_delete_someFC
DeleteDemuxAPI-Hasta: Removing flow cell run directory: /home/proj/stage/flowcells/novaseq/runs/test_delete_someFC
```

**Review by karlnyr — COMMENTED** [review]
This looks so nice!

I have one problem - there are occasions we could want to delete flow cells that actually do not have a demultiplexing out dir - nor a run dir - or vice versa. As the code is now we would fail to instantiate the API due to the glob missing - which is not what we would want, could you perhaps simply do a warning that we could not find the run dir/demultiplexing out dir instead?

**Comment by ChrOertlin** [comment]
> This looks so nice!
> 
> I have one problem - there are occasions we could want to delete flow cells that actually do not have a demultiplexing out dir - nor a run dir - or vice versa. As the code is now we would fail to instantiate the API due to the glob missing - which is not what we would want, could you perhaps simply do a warning that we could not find the run dir/demultiplexing out dir instead?

Yes, I noticed that too but was not aware that it could be like that. I now output a warning, instead of raising an error.

**Comment by ChrOertlin** [comment]
cg demultiplex delete-flow-cell -f someFC --run-dir
/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/sqlalchemy/orm/relationships.py:2231: SAWarning: Cascade settings "delete, merge, save-update" should not be combined with a viewonly=True relationship.   This configuration will raise an error in version 1.4.  Note that in versions prior to 1.4, these cascade settings may still produce a mutating effect even though this relationship is marked as viewonly=True.
  "viewonly=True." % (", ".join(sorted(non_viewonly)))
Running cg demultiplex.
Called undefined __fields__ on HousekeeperAPI, please wrap
Set environment to stage
Called undefined __fields__ on HousekeeperAPI, please wrap
Instantiating cgstats api
DeleteDemuxAPI: Path None not found for someFC
Are you sure you want to delete the flow cell from the following databases:
cg-stats=False
Demultiplexing-dir=False
Housekeeper=False
Init_files=False
Run-dir=True
Statusdb=False [y/N]: y
/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/sqlalchemy/orm/relationships.py:1998: SAWarning: Setting backref / back_populates on relationship Demux.samples to refer to viewonly relationship Sample.demuxes should include sync_backref=False set on the Demux.samples relationship.  (this warning may be suppressed after 10 occurrences)
  (rel_b, rel_a, rel_b),
/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/sqlalchemy/orm/relationships.py:1998: SAWarning: Setting backref / back_populates on relationship Sample.demuxes to refer to viewonly relationship Demux.samples should include sync_backref=False set on the Sample.demuxes relationship.  (this warning may be suppressed after 10 occurrences)
  (rel_b, rel_a, rel_b),
DeleteDemuxAPI-Hasta: No demultiplexing directory found: None
DeleteDemuxAPI-Hasta: Removing flow cell run directory: /home/proj/stage/flowcells/novaseq/runs/delete_someFC

**Comment by ChrOertlin** [comment]
cg demultiplex delete-flow-cell -f someFC --status-db
/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/sqlalchemy/orm/relationships.py:2231: SAWarning: Cascade settings "delete, merge, save-update" should not be combined with a viewonly=True relationship.   This configuration will raise an error in version 1.4.  Note that in versions prior to 1.4, these cascade settings may still produce a mutating effect even though this relationship is marked as viewonly=True.
  "viewonly=True." % (", ".join(sorted(non_viewonly)))
Running cg demultiplex.
Called undefined __fields__ on HousekeeperAPI, please wrap
Set environment to stage
Called undefined __fields__ on HousekeeperAPI, please wrap
Instantiating cgstats api
Are you sure you want to delete the flow cell from the following databases:
cg-stats=True
Demultiplexing-dir=True
Housekeeper=True
Init_files=True
Run-dir=True
Statusdb=True [y/N]: y
/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/sqlalchemy/orm/relationships.py:1998: SAWarning: Setting backref / back_populates on relationship Demux.samples to refer to viewonly relationship Sample.demuxes should include sync_backref=False set on the Demux.samples relationship.  (this warning may be suppressed after 10 occurrences)
  (rel_b, rel_a, rel_b),
/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/sqlalchemy/orm/relationships.py:1998: SAWarning: Setting backref / back_populates on relationship Sample.demuxes to refer to viewonly relationship Demux.samples should include sync_backref=False set on the Sample.demuxes relationship.  (this warning may be suppressed after 10 occurrences)
  (rel_b, rel_a, rel_b),
DeleteDemuxAPI-Hasta: Removing flow cell demultiplexing directory: /home/proj/stage/demultiplexed-runs/test_someFC
DeleteDemuxAPI-Hasta: Removing flow cell run directory: /home/proj/stage/flowcells/novaseq/runs/test_someFC
DeleteDemuxAPI-Housekeeper: Deleting sample sheet files with tag someFC
Fetching files with tags in [someFC,samplesheet]
DeleteDemuxAPI-Housekeeper: No files found with tag: someFC
DeleteDemuxAPI-Housekeeper: Deleting fastq and spring files related to flowcell someFC
DeleteDemuxAPI-StatusDB: Deleted flowcell someFC

**Comment by ChrOertlin** [comment]
stage:
Logging deploy ...
Getting deployer... done.
Getting last commit message and SHA... done.
Getting version of deploy scripts... /home/proj/stage/demultiplexed-runs
done.
Log deploy... done.
cg, version 38.0.0

prod:
Logging deploy ...
Getting deployer... done.
Getting last commit message and SHA... done.
Getting version of deploy scripts... /home/hiseq.clinical
done.
Log deploy... done.
cg, version 38.0.0

**Pull request merged**